### ISSUE TITLE
WASM runtime telemetry

### ISSUE THREAD

**Issue by jcetkov** [envoyproxy/envoy#20850]
*Description*:
Currently there is not possible to alert on (WASM) extension failure. 
It would be very desirable to  to allow monitoring of WASM extensions, particularly failing in `fail_open` state. 
(it would be nice to have _fail close_ as well; while that's inferable from no traffic, it would help faster diagnostics)

it would also be helpful to have indication of how many distinct filters are applied if practical (as I understand, currently there is a wasm runtime per worker thread and possibly per listener or filter chain, so it makes it difficult to observe if my extensions are applied as expected)

[optional *Relevant Links*:]
currently there is no stats regarding health of an extension
https://www.envoyproxy.io/docs/envoy/latest/configuration/other_features/wasm

cc: @PiotrSikora

**Comment by wbpcode** [comment]
cc @PiotrSikora

**Comment by bryanwux** [comment]
I agree this feature will be very helpful.

**Comment by PiotrSikora** [comment]
Yeah, I agree that we should add it. It might be a good starter issue. cc @daixiang0

**Comment by daixiang0** [comment]
@bryanwux I am busy with something else, do you want to pick it up?

**Comment by bryanwux** [comment]
Yes, I can pick up this issue.

**Comment by tsukiy0** [comment]
This would be a great feature to have.

@bryanwux have you made a start already? I'd be happy to work on this with you.

I don't have a lot of experience with C++ 😅 , but have started debugging the Envoy repo and found the [`Context:: failStream` method](https://github.com/envoyproxy/envoy/blob/518772982bff3d630306a26622f80d564d3b25ea/source/extensions/common/wasm/context.cc#L1595) to be a suitable place for incrementing a fail metric of some kind - it is called from the [host](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/a0db71f4f326867e5d1c2a77510e1cafb3bd49a9/src/context.cc#L34) whenever something goes wrong. In terms of what the metric would look like, I was thinking a counter called `wasm.<name>.fail`, where `<name>` is the name used to configure the WASM filter - then perhaps at a later point we could follow what [ext_authz](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter#statistics) did with the `stat_prefix` configuration to customize `<name>`.

@jcetkov @PiotrSikora @daixiang0 what are your thoughts, any guidance would be appreciated!
